src/metabase/pulse.clj

```diff
@@ -155,27 +155,39 @@
     (virtual-card-of-type? dashcard "link")
     (dashcard-link-card->content dashcard)
 
-    ;; text cards has existed for a while and I'm not sure if all existing text cards
+    ;; text cards have existed for a while and I'm not sure if all existing text cards
     ;; will have virtual_card.display = "text", so assume everything else is a text card
     :else
     (let [parameters (merge-default-values (params/parameters pulse dashboard))]
       (-> dashcard
           (params/process-virtual-dashcard parameters)
           :visualization_settings))))
 
+(defn- dashcards->content
+  [dashcards pulse dashboard]
+  (let [ordered-dashcards (sort dashcard-comparator dashcards)]
+    (doall (for [dashcard ordered-dashcards
+                 :let  [content (dashcard->content dashcard pulse dashboard)]
+                 :when (some? content)]
+             content))))
+
+(defn- tab->text
+  [{:keys [name]}]
+  {:slack {:text (str (format "# %s" name))}
+   :email {:text (str (format "# %s\n---" name))}})
```

Comment:
how would that change gonna be to be exact?

with this approach, we can reuse the rendering code for markdown which is quite nice.
and I also think having a way to provide platform-specific render could be beneficial in the future since the rendering code will call `recur`, so it could render both card, or text.

Reply:
Good question. For example, you could change the way the `pulse` is rendered to html in clojure before it gets to the mustache template, [here](https://github.com/metabase/metabase/blob/cce120266b926d9f95d5df3ec54ea695060665c7/src/metabase/email/messages.clj#L481). This is pseudocode, but you get the idea:
```clojure
        hiccup          (if has-tabs?
                          (into [:div]
                                (for [{:keys [tab-title cards]} tabs]
                                  (into [:div
                                         [:h1 (render-tab-title tab-title)]
                                         [:hr]
                                         (map :content (render-cards cards))])))
                          (into [:div] (map :content (render-cards (:cards only-tab))))
        message-body    (assoc message-context :pulse (html hiccup)
                               :filters filters
                               :iconCid (:content-id icon-attachment))
```
You could probably do this in the mustache template, but it's easier in clojure for us BE devs.

Another benefit of this approach is that it sets the stage for per-tab filters in the future, which is [planned](https://www.notion.so/metabase/Bring-tabbed-navigation-to-dashboards-dd2e71b4d1954f8188ecbd9914e81420?pvs=4#a163d59f531545b3a8b444f0425e877a).